### PR TITLE
rpi_pico: Added Generic PIO support and a PIO based UART driver

### PR DESCRIPTION
**Pull request by iocapa**
Needed some extra uarts some time ago (interrupt capable).

Knew that [#44316](https://github.com/zephyrproject-rtos/zephyr/pull/44316) existed, but I needed something fast and decided to roll my own.

@yonsch and everyone else.
Feedback is appreciated.

Cleaned the code up a bit and opened this PR, as it was sitting on my drive for some time.
Leaving this here, maybe someone could use it in their own project or maybe it would be helpful in current/future PIO implementations.

Some technical details:
Uses the existing `bus` / `on-bus` pattern. The PIO is the bus and manages resources / interrupts. The child devices implement the functionality.

Does not use the `pico sdk` or `pioasm`. This was intended, as the PIO is not really a complicated piece of hardware at the register level (program design is so-so). The programs are crafted by the peripheral implementing the function using whatever is fit for that implementation (there are some helper macros provided). This allows for more flexibility, for example, in the case of UART, a patch operation (dynamically changing instructions during runtime) is desired to change the number of stop bits.

Instruction / State machine allocation is performed at driver `init`. A static allocation scheme was experimented with in the beginning, but it falls short for devices sharing program space.

The rule on usage depends on the driver. The idea is that the child device should take care not to access anything (SMs / instructions / registers) not specifically allocated to it, including IRQ flags. For example, a driver using only SM0 should only use IRQ flags 0 and 4, a driver using all 4 SMs could use all flags.

Uart:
Not fully tested in all modes (data/stop lengths) but looks like it works as it should for shell / logging operations.
Experiments were done with adding TX parity suport / wide (9 bit) support, but I don't think it's useful ATM.

Pretty sure I omitted some technical details, but if anyone has any question, feel free to ask.

```
/ {
	chosen {
		zephyr,console = &pio0_uart0;
		zephyr,shell-uart = &pio0_uart0;
	};
};

&uart0 {
	status = "disabled";
};

&pio0 {
	status = "okay";

	pio0_uart0: pio0_uart0 {
		compatible = "raspberrypi,pico-pio-uart";

		status = "okay";

		pio,interrupts = <0>;
		pio,interrupt-names = "shared";

		pinctrl-0 = <&pio0_uart0_default>;
		pinctrl-names = "default";

		current-speed = <115200>;
	};
};
```
```
#include <zephyr/dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h>

&pinctrl {
	pio0_uart0_default: pio0_uart0_default {
		tx_gpio {
			pinmux = <PIO0_P0>;
		};

		rx_gpio {
			pinmux = <PIO0_P1>;
			input-enable;
			bias-pull-up;
		};
	};
};
```

```
*** Booting Zephyr OS build 1ac10e9d44d0 ***
[00:00:01.001,000] <inf> main: One second passed...
[00:00:02.001,000] <inf> main: One second passed...
[00:00:03.001,000] <inf> main: One second passed...
[00:00:04.001,000] <inf> main: One second passed...
uart:~$ help
Please press the <Tab> button to see all available commands.
You can also use the <Tab> button to prompt or auto-complete all commands or its subcommands.
You can try to call commands with <-h> or <--help> parameter for more information.

Shell supports following meta-keys:
  Ctrl + (a key from: abcdefklnpuw)
  Alt  + (a key from: bf)
Please refer to shell documentation for more details.

Available commands:
  clear    :Clear screen.
  device   :Device commands
  devmem   :Read/write physical memory
            Usage:
            Read memory at address with optional width:
            devmem address [width]
            Write memory at address with mandatory width and value:
            devmem address <width> <value>
  help     :Prints the help message.
  history  :Command history.
  kernel   :Kernel commands
  log      :Commands for controlling logger
  resize   :Console gets terminal screen size or assumes default in case the
            readout fails. It must be executed after each terminal width change
            to ensure correct text display.
  shell    :Useful, not Unix-like shell commands.
[00:00:05.002,000] <inf> main: One second passed...
[00:00:06.002,000] <inf> main: One second passed...
```

### CODE REVIEW

**Review by soburi — COMMENTED** [review]
> Instruction / State machine allocation is performed at driver init. A static allocation scheme was > experimented with in the beginning, but it falls short for devices sharing program space.

I think static allocation seems possible in theory. It would be helpful if you could give me more details about the problem.

File: include/zephyr/drivers/misc/rpi_pico_pio/rpi_pico_pio.h [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+	bool enabled;
+};
+
+/**
+ * @brief Register a shared IRQ
+ *
+ * @param dev Pointer to the parent device
+ * @param cfg Pointer to the IRQ configuration
+ */
+void rpi_pico_pio_irq_register(const struct device *dev, struct rpi_pico_pio_irq_cfg *cfg);
```
[the remark this answers is not in the document]
Comment (iocapa) [reply include/zephyr/drivers/misc/rpi_pico_pio/rpi_pico_pio.h:169]: @soburi  

Actually you can have as many devices as you want (can) on a PIO instance. See below for how I'm adding 4 extra uarts (2 per each PIO instance). In practice, you are only limited by the number of SMs / Instruction memory availability as there's no real inter dependency between SMs.

The 'bus' / 'on-bus' thing is mostly for static checking as you are required to be on a PIO parent device. This could be however implemented with a phandle property for each device, but it makes the DTS and checks more confusing.

The intermediate IRQ routing is done for potential load sharing purposes. You could for example overlay PIO IRQ0 to a higher prio and assign it to a child device that requires that higher prio, or you could have one or more child devices that need separate IRQ lines.

```
&pio0 {
	status = "okay";

	pio0_uart0: pio0_uart0 {
		compatible = "raspberrypi,pico-pio-uart";

		pio,interrupts = <0>;
		pio,interrupt-names = "shared";

		pinctrl-0 = <&pio0_uart0_default>;
		pinctrl-names = "default";

		current-speed = <115200>;
	};

	pio0_uart1: pio0_uart1 {
		compatible = "raspberrypi,pico-pio-uart";

		pio,interrupts = <1>;
		pio,interrupt-names = "shared";

		pinctrl-0 = <&pio0_uart1_default>;
		pinctrl-names = "default";

		current-speed = <115200>;
	};
};

&pio1 {
	status = "okay";

	pio1_uart0: pio1_uart0 {
		compatible = "raspberrypi,pico-pio-uart";

		pio,interrupts = <0>;
		pio,interrupt-names = "shared";

		pinctrl-0 = <&pio1_uart0_default>;
		pinctrl-names = "default";

		current-speed = <115200>;
	};

	pio1_uart1: pio1_uart1 {
		compatible = "raspberrypi,pico-pio-uart";

		pio,interrupts = <1>;
		pio,interrupt-names = "shared";

		pinctrl-0 = <&pio1_uart1_default>;
		pinctrl-names = "default";

		current-speed = <115200>;
	};
};
```

**Comment by iocapa** [comment]
> > Instruction / State machine allocation is performed at driver init. A static allocation scheme was > experimented with in the beginning, but it falls short for devices sharing program space.
> 
> I think static allocation seems possible in theory. It would be helpful if you could give me more details about the problem.

It's a bit of a hack, but the idea is as follows:
Each device has the default `sm-count` and `instr-count` properties in their bindings.
After that, you expand all all children and compute offsets based on the parent index and the sum of all props till the device index.
This works, but to share program space you would need to expand based on `compat`, and I'm not sure if this is possible at all.

Here is a snippet:
```
/**
 * @brief Compute resource offset based on other child nodes
 * 
 * Generates an expression in the following form:
 * ((N >= M) ? 0 : DT_PROP(res))
 * 
 * Will throw an error if other nodes do not have the 'res' property
 * This is required as a device not requiring the PIO should not be the
 * child of a PIO instance.
 * 
 * Where:
 * N           = Index of a node in the parent binding
 * M           = Index of the current node in the parent binding
 * res         = Value of the res property for the child_node
 */
#define RPI_PICO_PIO_COUNT_RES_OFFSET(child_node, node_id, res)                                        \
       ((DT_NODE_CHILD_IDX(child_node) >= DT_NODE_CHILD_IDX(node_id))                          \
               ? 0 : DT_PROP_OR(child_node, res, 0))

/**
 * @brief Gets one resource ID and offsets it
 * 
 * Generates a sum of offsets base on the 'res' property of all nodes:
 * 
 * ((N[0] >= M) ? 0 : res) + ((N[1] >= M) ? 0 : res) // ...
 */
#define RPI_PICO_PIO_GET_RES_ID(node_id, res, offset)                                          \
       ((DT_FOREACH_CHILD_STATUS_OKAY_SEP_VARGS(                                               \
               DT_PARENT(node_id), RPI_PICO_PIO_COUNT_RES_OFFSET, (+), node_id, res)           \
        ) + offset)

/**
 * @brief Assign a unique state machine ID from the available pool
 * 
 * State machines are allocated based on the DT structure.
 * Each peripheral implemented by PIO must have the 'sm_count'
 * property. This property must have a value equal to the number
 * of state machines required by the peripheral.
 * 
 * Example devicetree fragment:
 * 
 * @code{.dts}
 *     pio {
               status = "okay";
 *             c1: child-1 {
 *                     sm-count = <2>; // Fixed, based on peripheral requirements
 *             };
 *             c2: child-2 {
 *                     sm-count = <2>;
 *             };
 *     };
 * @endcode
 * 
 * Example bindings fragment:
 * 
 * @code{.yaml}
 *     include: "raspberrypi,pico-pio-device.yaml"
 *     properties:
 *       sm-count:
 *         type: int
 *         default: 2
 *         const: 2
 * @endcode
 * 
 * Example usage:
 * 
 * @code{.c}
 *     // child_2.c
 *     DT_RPI_PICO_PIO_ASSIGN_SM(node, 0)      // 2
 *     DT_RPI_PICO_PIO_ASSIGN_SM(node, 1)      // 3
 *
 *     // child_1.c
 *     DT_RPI_PICO_PIO_ASSIGN_SM(node, 0)      // 0
 *     DT_RPI_PICO_PIO_ASSIGN_SM(node, 1)      // 1
 * @endcode
 * 
 * The offset must be contained in the number of state machines assigned
 * to the peripheral. For example, requesting an ofsset of 2 for a sm-count
 * of 2, will result in an error.
 * 
 * @param node_id node identifier
 * @param offset The offset of the state machine in the assigned group
 * @return the assigned state machine's index
 */
#define DT_RPI_PICO_PIO_ASSIGN_SM(node_id, offset)                                             \
       RPI_PICO_PIO_GET_RES_ID(node_id, sm_count, offset)

/**
 * @brief Assign a unique state machine ID from the available pool
 * 
 * @param inst instance number
 * @param offset The offset of the state machine in the assigned group
 * @return the assigned state machine's index
 * 
 * @see DT_RPI_PICO_PIO_ASSIGN_SM
 */
#define DT_INST_RPI_PICO_PIO_ASSIGN_SM(inst, offset)                                           \
       DT_RPI_PICO_PIO_ASSIGN_SM(DT_DRV_INST(inst), offset)

/**
 * @brief Assign a unique program instruction ID from the available pool
 * 
 * Program instructions are allocated based on the DT structure.
 * Each peripheral implemented by PIO must have the 'sm_count'
 * property. This property must have a value equal to the number
 * of instructions required by the peripheral.
 * 
 * Example devicetree fragment:
 * 
 * @code{.dts}
 *     pio {
               status = "okay";
               clocks = <&system_clk>;
 *             c1: child-1 {
 *                     instr-count = <16>; // Fixed, based on peripheral requirements
 *             };
 *             c2: child-2 {
 *                     instr-count = <16>;
 *             };
 *     };
 * @endcode
 * 
 * Example usage:
 * 
 * @code{.c}
 *     // child_2.c
 *     DT_RPI_PICO_PIO_ASSIGN_INSTR(node, 0)   // 16
 *     DT_RPI_PICO_PIO_ASSIGN_INSTR(node, 8)   // 24
 *
 *     // child_1.c
 *     DT_RPI_PICO_PIO_ASSIGN_INSTR(node, 0)   // 0
 *     DT_RPI_PICO_PIO_ASSIGN_INSTR(node, 8)   // 8
 * @endcode
 * 
 * The offset must be contained in the number of program instructions assigned
 * to the peripheral. For example, requesting an ofsset of 16 for a instr-count
 * of 16, will result in an error.
 * 
 * @param node_id node identifier
 * @param offset The offset of the program instruction in the assigned group
 * @return the assigned program instruction's index
 */
#define DT_RPI_PICO_PIO_ASSIGN_INSTR(node_id, offset)                                          \
       RPI_PICO_PIO_GET_RES_ID(node_id, instr_count, offset)

/**
 * @brief Assign a unique instruction location from the available pool
 * 
 * @param inst instance number
 * @param offset The offset of the instruction in the assigned group
 * @return the assigned instructions's index
 * 
 * @see DT_RPI_PICO_PIO_ASSIGN_INSTR
 */
#define DT_INST_RPI_PICO_PIO_ASSIGN_INSTR(inst, offset)                                                \
       DT_RPI_PICO_PIO_ASSIGN_INSTR(DT_DRV_INST(inst), offset)
```

File: dts/bindings/serial/raspberrypi,pico-pio-uart.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+compatible: "raspberrypi,pico-pio-uart"
+
+include: [
+  "raspberrypi,pico-pio-device.yaml",
+  uart-controller.yaml,
+  pinctrl-device.yaml
+]
+
+properties:
```
Comment (galak) [inline dts/bindings/serial/raspberrypi,pico-pio-uart.yaml:14]: uart-controller.yaml now has `stop-bits` and `data-bits`, so you can remove them here

**Review by galak — CHANGES_REQUESTED** [review]
The interrupts for PIO should be done as a nested interrupt controller and thus their representation in devicetree would utilize normal interrupt properties.

**Comment by iocapa** [comment]
> The interrupts for PIO should be done as a nested interrupt controller and thus their representation in devicetree would utilize normal interrupt properties.

Open to suggestions, maybe I'm missing some things.
Thought about this initially, however, I think there are some issues in moving this to a 2nd level interrupt scheme.

* You still need to know the index of the PIO interrupt as you need to set one of 2 (future IP revisions could have more) internal PIO registers to enable/disable. This still requires some kind of mapping from the assigned IRQ id to the actual PIO IRQ id.

* Since lines are shared, you need to filter events inside the ISR, as all of them will be called if enabled, this has to be done by the  child device. Filtering could be done at the initial handler level, but it creates runtime dependencies since you need to know the SM indexes, and these are allocated at init.

* You could potentially have a maximum of 2*4 IRQs per PIO (for example, 4 devices). And, depending on use-case you can mix the assignments to satisfy some system requirements (maybe timing). I'm not really sure how this could be accomplished with a 2nd level INTC.

```
&pio0 {
	status = "okay";
	device0: device0 {
		compatible = "raspberrypi,pico-pio-future-device";
		pio,interrupts = <0>, <1>;
		pio,interrupt-names = "tx", "rx";
	};
	device1: device1 {
		compatible = "raspberrypi,pico-pio-future-device";
		pio,interrupts = <1>, <1>;
		pio,interrupt-names = "tx", "rx";
	};
        // .....
	deviceN: deviceN {
		compatible = "raspberrypi,pico-pio-future-device";
		pio,interrupts = <1>, <1>;
		pio,interrupt-names = "tx", "rx";
	};
```

File: dts/bindings/serial/raspberrypi,pico-pio-uart.yaml (hunk 2 above)
Comment (iocapa) [reply dts/bindings/serial/raspberrypi,pico-pio-uart.yaml]: Done

File: drivers/misc/CMakeLists.txt [hunk 3]
```diff
@@ -3,4 +3,4 @@
 add_subdirectory_ifdef(CONFIG_ARM_ETHOS_U ethos_u)
 add_subdirectory_ifdef(CONFIG_FT800 ft8xx)
 add_subdirectory_ifdef(CONFIG_GROVE_LCD_RGB grove_lcd_rgb)
-add_subdirectory_ifdef(CONFIG_PIO_RPI_PICO pio_rpi_pico)
+add_subdirectory_ifdef(CONFIG_RPI_PICO_PIO rpi_pico_pio)
```
Comment (soburi) [inline drivers/misc/CMakeLists.txt:6]: I don't think it's necessary to change the name.
It is generally named [function name]_[board name].
for example, "gpio_rpi_pico"

**Review by soburi — CHANGES_REQUESTED** [review]

File: drivers/misc/CMakeLists.txt (hunk 3 above)
Comment (iocapa) [reply drivers/misc/CMakeLists.txt:6]: Agree.
Will revisit sometime in the next few days.

File: boards/arm/rpi_pico/doc/index.rst [hunk 4]
```diff
@@ -83,7 +83,7 @@ hardware features:
      - :dtcompatible:`raspberrypi,pico-flash`
    * - UART (PIO)
      - :kconfig:option:`CONFIG_SERIAL`
-     - :dtcompatible:`raspberrypi,pico-uart-pio`
+     - :dtcompatible:`raspberrypi,pico-pio-uart`
```
Comment (soburi) [inline boards/arm/rpi_pico/doc/index.rst:86]: I think it's better to leave it as is.
The format [manufacturere],[device]-[implement-variant] is commonly used.

for example.
bindings/sensor/bosch,bme280-i2c.yaml
bindings/sensor/bosch,bme280-spi.yaml

**Review by soburi — CHANGES_REQUESTED** [review]

File: dts/bindings/misc/raspberrypi,pico-pio-device.yaml [hunk 5]
```diff
@@ -1,9 +1,17 @@
+# Copyright (c) 2023 Ionut Pavel <iocapa@iocapa.com>
 # Copyright (c) 2023, TOKITA Hiroshi
 # Copyright (c) 2023, Yonatan Schachter
 # SPDX-License-Identifier: Apache-2.0
 
 description: Raspberry Pi Pico PIO device
 
-compatible: "raspberrypi,pico-pio-device"
+on-bus: rpi-pico-pio
```
Comment (soburi) [inline dts/bindings/misc/raspberrypi,pico-pio-device.yaml:8]: In my understanding, this PR contains

1. Addition of interrupt handling
2. PIO program macro
3. Introduction of buses
4. Adding samples

I think 2 and 3 are easily acceptable, so you might want to raise them as separate PRs.
(In other words, 1 is probably  controversial.)

Comment (iocapa) [reply dts/bindings/misc/raspberrypi,pico-pio-device.yaml]: @soburi 

This PR contains a different implementation for handling the PIO / PIO devices.
**There is no API compatibility with the current implementation**, therefore it cannot be split, as builds will fail.

Aside from some similarities on the CMake/Kconfig/naming side, pretty much everything on the source side is a different design, see:

The "PIO" driver, contains (shared) instructions / SM allocators and shared interrupts handling.
[pio_rpi_pico.c](https://github.com/zephyrproject-rtos/zephyr/blob/293f573201218758d64b7952bc9a24752841794b/drivers/misc/pio_rpi_pico/pio_rpi_pico.c)
[pio_rpi_pico.h](https://github.com/zephyrproject-rtos/zephyr/blob/293f573201218758d64b7952bc9a24752841794b/include/zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h)

This is a full(ish) flegded uart over PIO implementation supporting multiple data/stop bits configurations and interrupt driven mode.
[uart_rpi_pico_pio.c](https://github.com/zephyrproject-rtos/zephyr/blob/293f573201218758d64b7952bc9a24752841794b/drivers/serial/uart_rpi_pico_pio.c)

**Comment by soburi** [comment]
@iocapa 

Let me confirm some premises.

What configuration of the drivers and instances is this PR trying to make possible?
(In my look, this PR considers dynamically sharing the PIO by two or more drivers.)

I think ...

PIO can only load one type of program, so multiple types of drivers cannot run simultaneously. (Multiple instances can up to state machine number.)
And the devicetree statically determines the configuration of the system.
Because of this constraint, I think PIO is occupied by one type of driver while the system is running. (If so, it seems better to implement interrupt handling in a separate driver.)

**Comment by soburi** [comment]
> The PIO can run multiple driver types in parallel.
> (E.g. if enough resources, you could have one UART one SPI and one WS2812 driver running on the same PIO instance)

Is it a case where a single PIO realizes multiple functions, for example, GPIO0 to UART-TX, GPIO1 to UART-RX, GPIO2 to SPI-CS... right? (It may be similar to USB MFD.)

In my opinion (Only just an idea now, not verified), in such cases, it's better to have each child node implement the interrupt handling, and the child nodes implement the individual features of the device, as an image of dts as follows.

```
&pio0 {
      dispatcher: pio_dispatcher {
           pio_uart: uart { ... };
           pio_spi: spi { ... };
           pio_ws2812: ws2812 { ... };
      };
};
```

The current PIO driver is just a placeholder, so
Child node drivers are free to implement. It can be implemented within Zephyr's general conventions. (It is important.)
Considering the possibility of various implementations, child nodes should be able to implement freely here. In other word, I want to the new API that is suggested in this PR can select by driver developer, not enforce to use it.

I think it is better to implement your interrupt dispatcher as a driver corresponding to the above dispatcher node and implement each driver as a child node. In this case, the implementation being done here is moved one level down in the dts tree. And the new API publish as dispatcher driver's API. PIO driver basically no need to modify.

There may are various opinions about the design, so I think there will be discussions.
In my opinion, It is preferable the new API should be "choosable" and give driver implementors more choice.

**Comment by soburi** [comment]
@iocapa 

I made a little code to demonstrate my idea.

https://github.com/soburi/zephyr/tree/pr56678_intr_dispatcher

In this code, I've renamed your code to make it an independent driver for pio_intr_dispatcher.
DTS puts this node under the pio node.
Since the PIO driver does nothing, child node drivers are free to use PIO as this driver does.

This way your code can be isolated as a module and reused when building other drivers.

**Comment by iocapa** [comment]
> In this code, I've renamed your code to make it an independent driver for pio_intr_dispatcher. DTS puts this node under the pio node. Since the PIO driver does nothing, child node drivers are free to use PIO as this driver does.
> 
> This way your code can be isolated as a module and reused when building other drivers.

@soburi 

The current PIO driver is the "API" used to implement child devices, it is possible to do this right now: 
```
&pio0 {
           pio_uart: uart { ... };
           pio_spi: spi { ... };
           pio_ws2812: ws2812 { ... };
};
```
I don't really understand what the extra "dispatcher" node is needed for. The top PIO driver allocates resources and dispatches callbacks to the shared IRQ lines. This is required since you cannot directly use physical IRQs from child devices as you will block other child devices from using that line (which is required for multiple child devices using the same line).

You cannot make the new API developer "choosable". Resource allocation needs to be managed from a single source (the PIO driver) or you'll end up with allocation conflicts.

To summarize:
* The top PIO driver is the "API", it manages resources and dispatches shared IRQ callbacks.
* There can be N child devices, and each of them can comply with pretty much any zephyr API convention (UART/SPI/I2C, etc). As long as it uses the PIO API and implementation conventions.
* You cannot mix the pico-sdk functions with the existing API as you'll have resource sharing conflicts.

**Comment by soburi** [comment]
> > In this code, I've renamed your code to make it an independent driver for pio_intr_dispatcher. DTS puts this node under the pio node. Since the PIO driver does nothing, child node drivers are free to use PIO as this driver does.
> > This way your code can be isolated as a module and reused when building other drivers.
> 
> @soburi
> 
> The current PIO driver is the "API" used to implement child devices, it is possible to do this right now:
> 
> ```
> &pio0 {
>            pio_uart: uart { ... };
>            pio_spi: spi { ... };
>            pio_ws2812: ws2812 { ... };
> };
> ```
> 
> I don't really understand what the extra "dispatcher" node is needed for. The top PIO driver allocates resources and dispatches callbacks to the shared IRQ lines. This is required since you cannot directly use physical IRQs from child devices as you will block other child devices from using that line (which is required for multiple child devices using the same line).
> 
> You cannot make the new API developer "choosable". Resource allocation needs to be managed from a single source (the PIO driver) or you'll end up with allocation conflicts.
> 
> To summarize:
> 
>     * The top PIO driver is the "API", it manages resources and dispatches shared IRQ callbacks.
> 
>     * There can be N child devices, and each of them can comply with pretty much any zephyr API convention (UART/SPI/I2C, etc). As long as it uses the PIO API and implementation conventions.
> 
>     * You cannot mix the pico-sdk functions with the existing API as you'll have resource sharing conflicts.

This described problem seems to be your subsystem-specific problem.
The PIO program handled by a single driver will not face this problem. (It no need to share irqs in this case).
So it is not a substantial reason to enforce to use non-standard interface because 'ALL' PIO drivers do not require it.
I think it not implements in the PIO driver, and it is should better to enclose it in the subsystem(=sub-node).

I agree that your program is helpful in many cases. But it is not a 'STANDARD' interface. I think it should not enforce unless all driver developers face the same problem.
(Of course, It is just my opinion. It should be better to discuss with many peoples.)

**Comment by iocapa** [comment]
> This described problem seems to be your subsystem-specific problem. The PIO program handled by a single driver will not face this problem. (It no need to share irqs in this case). So it is not a substantial reason to enforce to use non-standard interface because 'ALL' PIO drivers do not require it. I think it not implements in the PIO driver, and it is should better to enclose it in the subsystem(=sub-node).
> 
> I agree that your program is helpful in many cases. But it is not a 'STANDARD' interface. I think it should not enforce unless all PIO driver developers face the same problem. (Of course, It is just my opinion. It should be better to discuss with many peoples.)

Hey @soburi 

Let's take the following example:
We need 4 extra SPI controllers (doable with one pio instance).
Our DT will look something like this:
```
&pio0 {
	pio_spi0: spi0 { ... };
	pio_spi1: spi1 { ... };
	pio_spi2: spi2 { ... };
	pio_spi3: spi3 { ... };
};
```
Each SPI driver will use one SM and needs one interrupt routed to it.

Each PIO has 2 lines allocated to the NVIC (IRQ0 and IRQ1).

Each SM can trigger IRQ0 or IRQ1 with one of the following flags depending 
on SM configuration in `IRQ0_INTE` anr/or `IRQ1_INTE`.
```
SMx -> This can be triggered with `irq <irq_num> ( rel )`. Basically from a running program.
SMx_TXNFULL -> Triggered while Tx FIFO is not full.
SMx_RXNEMPTY -> Triggered while Rx FIFO is not empty.
```
Basically a big 12 input OR gate (4 SMs * 3 Events) for each IRQn line

Now, this is not possible with only 2 IRQ lines since to implement 4 instances
we would need 4 separate IRQ lines. This is why IRQ sharing is required. Otherwise
you would limit the maximum number of devices on a PIO to 2 (or 1 if a specific driver
requires 2 separate lines).

This PR uses a shared interrupt mechanism to basically enable each driver 
to register itself with a callback to one of the 2 interrupt lines.
The driver needs to check what flags were active to see if it is the one triggering
the interrupt. This is required since the ISR does not know anything about the source (due to the big OR gate).
This is similar to `shared-irq`, but implemented at the PIO level. This is required since you need to
translate an IRQn (index) into a NVIC index at runtime, and the PIO holds this routing information.

Looks like IRQ sharing is not a PIO specific problem and it's present in other (few) zephyr drivers.

So, IMO, interrupt sharing is mandatory unless we want to limit the capabilities of the PIO. And
I don't think this is a good ideea.

@galak @nashif @carlescufi @yonsch @tejlmand @dcpleung @gmarull 
Any comment on this one?

Regards,
Ionut

**Review by dcpleung — COMMENTED** [review]
Have you looked at `drivers/interrupt_controller/intc_shared_irq.c` for shared IRQ? There is no need to reinvent it if that works for you.

**Comment by iocapa** [comment]
> Have you looked at `drivers/interrupt_controller/intc_shared_irq.c` for shared IRQ? There is no need to reinvent it if that works for you.

@dcpleung
Yes, `shared-irq` (intc_share_irq.c) was considered. However. I couldn't find a way to make it work in this particular case.
The problem is that a translation from an NVIC index to a PIO index is required since the device implemented over PIO needs to know in what register to enable the interrupt (INT0, INT1). I see no way to extract this information from `shared-irq`.
That's the reason the `pio,interrupts` property exists at the node level.

**Comment by iocapa** [comment]
Hey @yonsch @soburi @galak @dcpleung. Can you have another look at this one?

I think the major point of debate around this was about interrupt handling.
I explained in the comments above why I feel this is necessary and why IMO there is no way around it.
Any other ideas are welcomed.

Another topic was the use of the raspberry pi API.
While I have nothing against it, again, IMO, it creates more issues than it solves. It has a lot of convoluted APIs for a pretty simple peripheral at the register level. Also future issues could be introduced by rpi SDK updates.

The PIO has 6 registers for each SM, and 3 of them (CLKDIV, ADDR, INSTR) are pretty self explanatory, the others (EXECCTRL, SHIFTCTRL, PINCTRL) are pretty well documented in the DS. So it's pretty easy to understand and configure. There are also other issues described in the conversations above.

Now, architecturally defining a PIO peripheral might not be that easy, but this is true also by using the SDK. Programs and flows still need to be thought of, and there has to be a pretty good understanding on how the PIO works (the DS is pretty good on this side).

Rpi provided examples are a good starting point, but they are pretty limited in scope and most of the times expanding them to a full featured peripheral requires a major rewrite.

If anyone has any questions/suggestions, please let me know.

P.S. The proposed API (including the UART drivers) was pretty well tested for some time in a limited production environment. No obvious issues till now. Also, I've used this api for another PIO based driver (JTAG/SWD) with quite interesting results.

Ionut

**Comment by iocapa** [comment]
> Is it possible to solve this problem by extending shared_irq?
> (I think this problem is essentially not PIO specific problem. It should abstract the problem and apply it to generic cases.)

@soburi `shared-irq` could technically be adapted to support pio shared IRQs, but it would mean adding PIO only code inside it, which would break it's generic approach IMO.

**Comment by soburi** [comment]
@yonsch Could you tell us your opinion, please?

**Comment by soburi** [comment]
Another idea is overwriting the "compatible" line by overlay file to replace the PIO driver completely.
We can think of PIO as a separate device for each loaded program. From this point of view, it can be a method suitable for Zephyr design.

**Comment by iocapa** [comment]
> Another idea is overwriting the "compatible" line by overlay file to replace the PIO driver completely. We can think of PIO as a separate device for each loaded program. From this point of view, it can be a method suitable for Zephyr design.

@soburi
Not sure I understand your idea. Can you give more details?
Currently, multiple devices can share the same program. This has a lot of benefits for many use-cases.

@all
Since the PIO SPI driver was committed as part of:
https://github.com/zephyrproject-rtos/zephyr/commit/85cbc7a96e1d032cb24f5a63ac9b6dbec61a122f
Compliance checks will fail as the commit is not compatible with this PR.

Most likely, to make this PR pass all checks again and maintain the SPI driver, me (or someone else) needs to recreate the driver using the framework in this PR (either by a new commit on this PR or chaining another PR starting from this one as base and temporary disabling the existing SPI driver).

It's not a massive amount of work, but it's work nonetheless.

So, I would appreciate some feedback on this PR to see where it's going. As more PIO based drivers will be created this will become harder to keep up to date.

**Comment by yonsch** [comment]
@iocapa @soburi
Sorry for taking so long to get to this PR.
As far as I can tell (tell me if I missed anything), it consists of the following:
1. Reimplementing the PIO device without pico-sdk
2. Adding an "assembler" implemented with macros
3. Adding support for interrupts
4. Adding more UART functionality
5. Making the pico-device binding a bus rather than a compatible
6. Converting the uart_pio sample to a shell
7. Renaming stuff (e.g. rpi pico pio to pio rpi pico)

Many of which are a topic for a discussion.

#### Reimplementing the PIO device without pico-sdk
This is a dilemma we're running into with most of the pico drivers. pico-sdk sometimes does pretty annoying things internally, which forces us to reimplement low level code. I generally think we should start migrating away from the pico-sdk code, but this is a big decision and requires a more thorough discussion.

#### Adding an "assembler" implemented with macros
This is a cool idea, but I wonder why we'd prefer it over using pioasm. I know that currently it's not available, but supposedly we'll one day have it as an installable package (https://github.com/raspberrypi/pico-sdk/issues/1345), which we could then use to assemble the programs on every build. It might require some more work around the build system, but that's less work than maintaining an assembler. Just think of debugging a PIO program and discovering that it's a "compiler" bug...

#### Adding support for interrupts
This should be very carefully planned, both on the PIO side and core side. I really should dive deeper into this before commenting further.

#### 4, 5 and 6
Sounds good overall

#### Renaming stuff (e.g. rpi pico pio to pio rpi pico)
Not sure if this was intentional or just an artifact. If it's intentional, and you have good reasoning to change a name, we can probably still change it. If it's just a matter of taste, I prefer to leave it as is.

**Comment by iocapa** [comment]
> #### Reimplementing the PIO device without pico-sdk
At least on the PIO side, the sdk is pretty messy. Sure it's nice to read a few examples and make a really limited implementation for a peripheral. This falls apart pretty fast when you get into the more advanced PIO usage scenarios (interrupts, DMA, dynamic patching, execution from FIFO streams, dynamic reconfiguration, etc...).
The PIO is a pretty capable accelerator, it can do some really nice things, the SDK doesn't really help with that.

> #### Adding an "assembler" implemented with macros
The macros are not meant to replace pioasm, they're just to avoid random hardcoded hex numbers in source files. I have no strong feeling on any approach. The idea was to make it clear (and easily changeable) without the need to propagate *.pio files or the assembler. For example, there are some cases when you need to inject PIO instruction through the FIFOs, and this: `pio->fifo = PIO_ASM_IRQ(0, 1, PIO_ASM_INDEX(true, 0), PIO_SS(0, 0, 0))` looks better than `pio->fifo = 0xBEEF`.

> #### Adding support for interrupts
There are quite a lot of comments in this PR that explain why this cannot be avoided if maximum PIO flexibility is desired.

> #### Renaming stuff (e.g. rpi pico pio to pio rpi pico)
Files were not renamed from the names you had in your original PR (only initially, renaming was reverted quite some time ago).

**Comment by soburi** [comment]
Regarding the issue of IRQ dispatching, it seems necessary to see the progress of the discussion in #66505.
(I don't think it should be considered as a problem only to the PIO.)

As @yonsch shows, I think this PR can split into some parts. We can discuss start with 'light' issues.
